### ISSUE TITLE
Request: Add HD texture pack dumping/loading

### ISSUE THREAD

**Issue by Dilapidated** [snes9xgit/snes9x#474]
Would it be possible for SNE9x to add capabilities similar to Mesen/Dolphin that would allow dumping of SNES sprites, backgrounds, layers, etc for editing and reinsertion? I'm assuming outside of technical reasons this may not have been implemented due to how long it takes to re-texture games but with the current pace of AI neural networks entire SNES games could be re-textured fairly quickly now days.

**Comment by qwertymodo** [comment]
The SNES doesn't have any concept of textures like the GC/Wii does, everything is built from individual tiles that can be in a handful of different color formats, and the actual colors themselves are also completely separate.  If you try to create "HD" tiles individually, without knowing all of the possible ways they can be combined, the results look pretty terrible.  Just look at the Steam release of Chrono Trigger and how terrible the backgrounds looked in early versions, due to scaling that didn't respect the tiling.  There have been a few examples like an HD retexture of the original Metroid that turned out alright, but that's basically only possible because the tiles in that game are more or less designed to emphasize the grid on which they are placed.  And on top of that, even if you could come up with a texture pack that looked good, it would require modifying the DMA controller in ways that would likely cause all kinds of problems.  As much as I may like the idea in theory, I just don't see it happening.

**Comment by Dilapidated** [comment]
@qwertymodo Thanks for the quick response and for explaining it well. You make some very valid points regarding tiles and grids being issues. Perhaps these could be addressed somehow with the proper logic to handle them but I'm sure it'd be a lot of work. The idea of remastering some of these old SNES games is a very fascinating topic. Hopefully one day it'll be possible.

**Comment by bearoso** [comment]
I don't think actually replacing them in output would be difficult. You'd just checksum the final tile data and draw from a mapped replacement to a separate high resolution image during rasterization. As qwertymodo says, though, the number of tile combinations is going to be an order of magnitude higher on the SNES, which makes it infeasible, especially for the games that you'd want to do this for.

**Comment by Dilapidated** [comment]
It's encouraging to know dumping and replacing the output wouldn't be as challenging as initially thought. It makes sense that tile combinations and other issues would make such a retexture job doable but very painful to accomplish. 

Let's suppose the feature gets implemented. How difficult would it be to write logic to simplify the process of tiles and tile combinations outputs? For example something similar to the mesen link I posted above that takes these difficulties into consideration and outputs them in a more user friendly way based on configuration options selected? If the output data could constructed in such a way perhaps this idea could take off?

**Comment by qwertymodo** [comment]
You would also need to account for the fact that a lot of games achieve special effects by changing the color palettes without changing the tiles themselves. Often this is used to achieve rainbow flashing effects like the speed boost/shinespark in Super Metroid (also the Hyper Beam), but also in more subtle ways like water sparkling effects in Chrono Trigger.  Either the HD tiles would need to remain tied to the internal CGRAM or you'd need each texture to be tied to a tile *and* palette.

**Comment by bearoso** [comment]
Yeah, you’d either have to record every palette/tile combo as a bitmap or the tile as a bitmap and every palette ever applied to it. The former would yield a lot of things to change, but the latter would limit color depth. Things like fading effects would trigger a lot of output.

**Comment by Dilapidated** [comment]
I really appreciate you guys explaining the technical obstacles involved to make this happen instead of just dismissing my feature proposal, thank you.

So based on the conversations above I take from it that this would actually be feasible to implement but would require a lot logic and time to get it right. There's lots of variables to consider but in the end hopefully I've intrigued you all enough to consider giving it an attempt.

**Comment by Kolano** [comment]
It doesn't seem color palette and high definition are really related. Presumably if you provided tiles using the same (8-bit?) palette things could work out even with palette shifting. I understand some titles may do some fancier tile replacements/effects that may be difficult to account for, but I'd guess many titles wouldn't even need to account for specialized palettes and could just use straight replacements.

**Comment by bearoso** [comment]
If you limit the support to the same color palette and strictly increase resolution, you don't have to worry about it. Mesen let's artists use *more* colors, which if desired would require the artist to compensate by drawing all the palette states as well. Either way, you have to save the palettes (at least one) so they know how the image will look.

**Comment by Ifinxx** [comment]
I've been searching everywhere for a SNES emulator with the ability to change textures. This would be amazing! I really hope to see this implemented in the near future.

**Comment by DarkSagaBR** [comment]
SIgn me up, got everything set to make all spie games surpassing hd in no time... Jut point me to sch dumping load texture tool and say when.... a tip, apparently projectn6dumps sprites from MK trilogy, still I was unable to load, and even if does, it's too limited... Snes would be the thing

**Comment by masuta16** [comment]
yes it's possible

**Comment by diegofps** [comment]
Hi,

I've been thinking about this problem for the past few years as well, but I didn't limit it to a texture problem. I looked at it as "How to update these low-resolution games and bring them to 4K resolution?". I was checking the snes9x source when I found this post. Assuming we could use AI to achieve this, I identified three possible strategies:

1. **Make an output filter that receives the original frame and applies a set of transformations to upscale the images -** I investigated this one a little, collected a few frames, and applied GANs trained for super-resolution with a few filters. The problem here is that these are heavy operations and require large GPUs. The good thing is that as we are processing the entire frame, it does not break the tiles matching each other. It could even add extra effects.
2. **Make a single model that receives the original frame and produces the upscaled version -** The idea is to use a single neural network trained on upscaled images generated by the previous strategy. The downside is that we need to play the entire games in the original resolution, capture their frames, transform a large set of them using the process mentioned before, and train a large network on them. This would be much faster and possibly be executed in real time. However, the entire process is quite long and expensive.
3. **Upscale the tiles and update the emulator to use this enhanced image pack -** This is the strategy you've been discussing. The emulator would need to be adapted to draw to the original buffer and the enhanced buffer. I say draw the original buffer because I imagined that a few games might do some type of logic depending on pixel values (just guessing). I would also map each pair of tileset with a color palette and generate its upscaled version. The issue here is that there could be a lot of color palettes. Upscaling only the original image is not an option as the upscaled colors would not be easily matched to the color palette. However, we could likely keep the tiles matching if we: (1) extract the tileset; (2) rearrange the tiles to the order they will appear, next to each other; (3) upscale the rearranged tiles; and (4) undo tiles rearrangement to their original order. For those who don't match, we could add custom processes in the pipeline, like blending two of them with mirrored alpha in the edges to create a compatible one.

I spent some time working on Strategy 1. Here is a sample of Super Metroid generated by my pipeline to test the possibilities:

![1510268480014](https://github.com/snes9xgit/snes9x/assets/393781/41a4f728-3842-4f6f-bbf3-1f4d6dc53c68)
![1510269381641](https://github.com/snes9xgit/snes9x/assets/393781/7ac8a853-c324-4078-adb1-22ff849cf4fd)
![546931](https://github.com/snes9xgit/snes9x/assets/393781/ba6a520c-c893-4364-b76a-4f170e1dce6c)
![1510181361254](https://github.com/snes9xgit/snes9x/assets/393781/72f7be4e-ee13-4eba-8093-47e6906d927b)

I also spent some time on Strategy 2 but I didn't have enough data or computational power for it. I believe Strategy 3 is the closest one right now.

Keep in mind that this is just an experiment and that I was not using the most advanced models, or super-resolution models created specifically for pixel art games. Besides, we are starting to see Multimodal Models and it will become significantly easier to create high-res assets in the near future. Moreover, people could also use a different style, like making it look cartoonish, Van Gogh style, real-life, and so on. Despite the difficulties, these initial results made me sure that this would be an amazing feature to have.

**Comment by masuta16** [comment]
You can add new graphics over old image graphics

**Comment by diegofps** [comment]
Yes, but replacing will not work when the sizes are different, unfortunately. The problem is that snes9x's graphics engine is hard-coded for the original tile sizes. From what I could see in tileimpl.h, it goes as far as drawing the pixels manually and iterating over fixed sizes, like looping over 8-pixel lines. This is great for replicating the original SNES behavior but makes it harder to add such new features.

To achieve this, I believe we would need to redesign the graphics engine, ideally using OpenGL and GLSL to draw the textures more efficiently. Such redesign could possibly facilitate support for HD Mode 7 and similars (https://github.com/snes9xgit/snes9x/issues/529). If possible at all, it would require patience to port the original SNES way of thinking to this new format. Maybe it would even enhance the original rotation effect we see in Super Metroid during the space colony escape. This is how it looks in my current pipeline. As it is applied to the original frame, it inherits its discontinuous lines:

![1507845454036](https://github.com/snes9xgit/snes9x/assets/393781/7f891089-7973-4ad1-8d67-ca8d3315e5f8)

**Comment by masuta16** [comment]
I've heard about people using a graphic pack with hires images as a improvement in 4k on snes9x but I have no idea how to create it, but it's a very old implementation

**Comment by diegofps** [comment]
It would be interesting to take a look at it, even though they probably made just a proof of concept.

**Comment by dginolerasta-creator** [comment]
Hi everyone, I would be happy to participate to a project to make HD remastered version of DKC or DBZ 2 or other legendary snes games.

**Comment by diegofps** [comment]
Hi @dginolerasta-creator Maybe someone else can comment on the current state of HD features in the snes9x project, if it exists (I am not a contributor), but I heard that the zsnes team is working on a "super zsnes" that will support HD textures and audios. I didn't check it yet, but I would expect you could develop enhanced HD tiles and audios that could be exported to both emulators (once they both support it).

**Comment by dginolerasta-creator** [comment]
I would be delighted if such projects could come to fruition one day, and I am ready to contribute to making them happen. But I don't know how to find out if there are already projects underway or if there are people who would be interested in forming a team.
I actually posted this message following the release of the ZSNES new version, which has given things a new lease of life.
